# Problem statement
classify the Size_Categorie using SVM

month	month of the year: 'jan' to 'dec'
day	day of the week: 'mon' to 'sun'
FFMC	FFMC index from the FWI system: 18.7 to 96.20
DMC	DMC index from the FWI system: 1.1 to 291.3
DC	DC index from the FWI system: 7.9 to 860.6
ISI	ISI index from the FWI system: 0.0 to 56.10
temp	temperature in Celsius degrees: 2.2 to 33.30
RH	relative humidity in %: 15.0 to 100
wind	wind speed in km/h: 0.40 to 9.40
rain	outside rain in mm/m2 : 0.0 to 6.4
Size_Categorie 	the burned area of the forest ( Small , Large)

In [3]:
#import libraries
import pandas as pd 
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score, confusion_matrix
import warnings
warnings.filterwarnings('ignore')

# Importing CSV File

In [4]:
#reading a CSV file
df = pd.read_csv("forestfires.csv")

In [5]:
df.head()

month  day  FFMC   DMC     DC  ISI  temp  RH  wind  rain  ...  monthfeb  \
0   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0  ...         0   
1   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0  ...         0   
2   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0  ...         0   
3   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2  ...         0   
4   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0  ...         0   

   monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0         0         0         0         1         0         0         0   
1         0         0         0         0         0         0         1   
2         0         0         0         0         0         0         1   
3         0         0         0         1         0         0         0   
4         0         0         0         1         0         0         0   

   monthsep  size_category  
0         0          small  
1         0          small  
2         0          small  
3         0          small  
4         0          small  

[5 rows x 31 columns]

# EDA

In [8]:
#computes descriptive statistics of the DataFrame 
df.describe()

FFMC         DMC          DC         ISI        temp          RH  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean    90.644681  110.872340  547.940039    9.021663   18.889168   44.288201   
std      5.520111   64.046482  248.066192    4.559477    5.806625   16.317469   
min     18.700000    1.100000    7.900000    0.000000    2.200000   15.000000   
25%     90.200000   68.600000  437.700000    6.500000   15.500000   33.000000   
50%     91.600000  108.300000  664.200000    8.400000   19.300000   42.000000   
75%     92.900000  142.400000  713.900000   10.800000   22.800000   53.000000   
max     96.200000  291.300000  860.600000   56.100000   33.300000  100.000000   

             wind        rain         area      dayfri  ...    monthdec  \
count  517.000000  517.000000   517.000000  517.000000  ...  517.000000   
mean     4.017602    0.021663    12.847292    0.164410  ...    0.017408   
std      1.791653    0.295959    63.655818    0.371006  ...    0.130913   
min      0.400000    0.000000     0.000000    0.000000  ...    0.000000   
25%      2.700000    0.000000     0.000000    0.000000  ...    0.000000   
50%      4.000000    0.000000     0.520000    0.000000  ...    0.000000   
75%      4.900000    0.000000     6.570000    0.000000  ...    0.000000   
max      9.400000    6.400000  1090.840000    1.000000  ...    1.000000   

         monthfeb    monthjan    monthjul    monthjun    monthmar    monthmay  \
count  517.000000  517.000000  517.000000  517.000000  517.000000  517.000000   
mean     0.038685    0.003868    0.061896    0.032882    0.104449    0.003868   
std      0.193029    0.062137    0.241199    0.178500    0.306138    0.062137   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

         monthnov    monthoct    monthsep  
count  517.000000  517.000000  517.000000  
mean     0.001934    0.029014    0.332689  
std      0.043980    0.168007    0.471632  
min      0.000000    0.000000    0.000000  
25%      0.000000    0.000000    0.000000  
50%      0.000000    0.000000    0.000000  
75%      0.000000    0.000000    1.000000  
max      1.000000    1.000000    1.000000  

[8 rows x 28 columns]

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 31 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   month          517 non-null    object 
 1   day            517 non-null    object 
 2   FFMC           517 non-null    float64
 3   DMC            517 non-null    float64
 4   DC             517 non-null    float64
 5   ISI            517 non-null    float64
 6   temp           517 non-null    float64
 7   RH             517 non-null    int64  
 8   wind           517 non-null    float64
 9   rain           517 non-null    float64
 10  area           517 non-null    float64
 11  dayfri         517 non-null    int64  
 12  daymon         517 non-null    int64  
 13  daysat         517 non-null    int64  
 14  daysun         517 non-null    int64  
 15  daythu         517 non-null    int64  
 16  daytue         517 non-null    int64  
 17  daywed         517 non-null    int64  
 18  monthapr  

In [17]:
df.shape

(517, 29)

In [11]:
#Dropping the month and day columns
#month and day splitted and added in the data already so not useful columns
df.drop(["month","day"],axis=1,inplace =True)

In [13]:
df.head()

FFMC   DMC     DC  ISI  temp  RH  wind  rain  area  dayfri  ...  monthfeb  \
0  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0       1  ...         0   
1  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0       0  ...         0   
2  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0       0  ...         0   
3  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0       1  ...         0   
4  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0       0  ...         0   

   monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0         0         0         0         1         0         0         0   
1         0         0         0         0         0         0         1   
2         0         0         0         0         0         0         1   
3         0         0         0         1         0         0         0   
4         0         0         0         1         0         0         0   

   monthsep  size_category  
0         0          small  
1         0          small  
2         0          small  
3         0          small  
4         0          small  

[5 rows x 29 columns]

In [15]:
#getting duplicated columns
df.duplicated().sum()

8

In [20]:
#dropping the duplicated rows
df=df.drop_duplicates()
df.shape

(509, 29)

In [21]:
df['size_category'].value_counts()

small    371
large    138
Name: size_category, dtype: int64

# feature engineering
Changing large to '1' and small to '0'

In [22]:
df['size_category'] = np.where(df['size_category'].str.contains("large"), 1, 0)
df.head()

FFMC   DMC     DC  ISI  temp  RH  wind  rain  area  dayfri  ...  monthfeb  \
0  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   0.0       1  ...         0   
1  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   0.0       0  ...         0   
2  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   0.0       0  ...         0   
3  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   0.0       1  ...         0   
4  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   0.0       0  ...         0   

   monthjan  monthjul  monthjun  monthmar  monthmay  monthnov  monthoct  \
0         0         0         0         1         0         0         0   
1         0         0         0         0         0         0         1   
2         0         0         0         0         0         0         1   
3         0         0         0         1         0         0         0   
4         0         0         0         1         0         0         0   

   monthsep  size_category  
0         0              0  
1         0              0  
2         0              0  
3         0              0  
4         0              0  

[5 rows x 29 columns]

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 509 entries, 0 to 516
Data columns (total 29 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   FFMC           509 non-null    float64
 1   DMC            509 non-null    float64
 2   DC             509 non-null    float64
 3   ISI            509 non-null    float64
 4   temp           509 non-null    float64
 5   RH             509 non-null    int64  
 6   wind           509 non-null    float64
 7   rain           509 non-null    float64
 8   area           509 non-null    float64
 9   dayfri         509 non-null    int64  
 10  daymon         509 non-null    int64  
 11  daysat         509 non-null    int64  
 12  daysun         509 non-null    int64  
 13  daythu         509 non-null    int64  
 14  daytue         509 non-null    int64  
 15  daywed         509 non-null    int64  
 16  monthapr       509 non-null    int64  
 17  monthaug       509 non-null    int64  
 18  monthdec  

In [24]:
#separating the predictors (X) and the target variable (y)
X= df.iloc[:,0:28]
y = df.iloc[:,28]

In [25]:
X

FFMC    DMC     DC   ISI  temp  RH  wind  rain   area  dayfri  ...  \
0    86.2   26.2   94.3   5.1   8.2  51   6.7   0.0   0.00       1  ...   
1    90.6   35.4  669.1   6.7  18.0  33   0.9   0.0   0.00       0  ...   
2    90.6   43.7  686.9   6.7  14.6  33   1.3   0.0   0.00       0  ...   
3    91.7   33.3   77.5   9.0   8.3  97   4.0   0.2   0.00       1  ...   
4    89.3   51.3  102.2   9.6  11.4  99   1.8   0.0   0.00       0  ...   
..    ...    ...    ...   ...   ...  ..   ...   ...    ...     ...  ...   
512  81.6   56.7  665.6   1.9  27.8  32   2.7   0.0   6.44       0  ...   
513  81.6   56.7  665.6   1.9  21.9  71   5.8   0.0  54.29       0  ...   
514  81.6   56.7  665.6   1.9  21.2  70   6.7   0.0  11.16       0  ...   
515  94.4  146.0  614.7  11.3  25.6  42   4.0   0.0   0.00       0  ...   
516  79.5    3.0  106.7   1.1  11.8  31   4.5   0.0   0.00       0  ...   

     monthdec  monthfeb  monthjan  monthjul  monthjun  monthmar  monthmay  \
0           0         0         0         0         0         1         0   
1           0         0         0         0         0         0         0   
2           0         0         0         0         0         0         0   
3           0         0         0         0         0         1         0   
4           0         0         0         0         0         1         0   
..        ...       ...       ...       ...       ...       ...       ...   
512         0         0         0         0         0         0         0   
513         0         0         0         0         0         0         0   
514         0         0         0         0         0         0         0   
515         0         0         0         0         0         0         0   
516         0         0         0         0         0         0         0   

     monthnov  monthoct  monthsep  
0           0         0         0  
1           0         1         0  
2           0         1         0  
3           0         0         0  
4           0         0         0  
..        ...       ...       ...  
512         0         0         0  
513         0         0         0  
514         0         0         0  
515         0         0         0  
516         1         0         0  

[509 rows x 28 columns]

In [26]:
y

0      0
1      0
2      0
3      0
4      0
      ..
512    1
513    1
514    1
515    0
516    0
Name: size_category, Length: 509, dtype: int32

# inference
a series or column called "size_category" with 509 entries. The values in this column represent different categories of sizes, such as "small" as 0 and "large" as 1.

In [27]:
##Normalising the data as there is scale difference
def norm_func(i):
    x= (i-i.min())/(i.max()-i.min())
    return (x)

In [28]:
#transforming the data 
fires = norm_func(X)

In [29]:
fires

FFMC       DMC        DC       ISI      temp        RH      wind  \
0    0.870968  0.086492  0.101325  0.090909  0.192926  0.423529  0.700000   
1    0.927742  0.118194  0.775419  0.119430  0.508039  0.211765  0.055556   
2    0.927742  0.146795  0.796294  0.119430  0.398714  0.211765  0.100000   
3    0.941935  0.110958  0.081623  0.160428  0.196141  0.964706  0.400000   
4    0.910968  0.172984  0.110590  0.171123  0.295820  0.988235  0.155556   
..        ...       ...       ...       ...       ...       ...       ...   
512  0.811613  0.191592  0.771315  0.033868  0.823151  0.200000  0.255556   
513  0.811613  0.191592  0.771315  0.033868  0.633441  0.658824  0.600000   
514  0.811613  0.191592  0.771315  0.033868  0.610932  0.647059  0.700000   
515  0.976774  0.499311  0.711622  0.201426  0.752412  0.317647  0.400000   
516  0.784516  0.006547  0.115867  0.019608  0.308682  0.188235  0.455556   

        rain      area  dayfri  ...  monthdec  monthfeb  monthjan  monthjul  \
0    0.00000  0.000000     1.0  ...       0.0       0.0       0.0       0.0   
1    0.00000  0.000000     0.0  ...       0.0       0.0       0.0       0.0   
2    0.00000  0.000000     0.0  ...       0.0       0.0       0.0       0.0   
3    0.03125  0.000000     1.0  ...       0.0       0.0       0.0       0.0   
4    0.00000  0.000000     0.0  ...       0.0       0.0       0.0       0.0   
..       ...       ...     ...  ...       ...       ...       ...       ...   
512  0.00000  0.005904     0.0  ...       0.0       0.0       0.0       0.0   
513  0.00000  0.049769     0.0  ...       0.0       0.0       0.0       0.0   
514  0.00000  0.010231     0.0  ...       0.0       0.0       0.0       0.0   
515  0.00000  0.000000     0.0  ...       0.0       0.0       0.0       0.0   
516  0.00000  0.000000     0.0  ...       0.0       0.0       0.0       0.0   

     monthjun  monthmar  monthmay  monthnov  monthoct  monthsep  
0         0.0       1.0       0.0       0.0       0.0       0.0  
1         0.0       0.0       0.0       0.0       1.0       0.0  
2         0.0       0.0       0.0       0.0       1.0       0.0  
3         0.0       1.0       0.0       0.0       0.0       0.0  
4         0.0       1.0       0.0       0.0       0.0       0.0  
..        ...       ...       ...       ...       ...       ...  
512       0.0       0.0       0.0       0.0       0.0       0.0  
513       0.0       0.0       0.0       0.0       0.0       0.0  
514       0.0       0.0       0.0       0.0       0.0       0.0  
515       0.0       0.0       0.0       0.0       0.0       0.0  
516       0.0       0.0       0.0       1.0       0.0       0.0  

[509 rows x 28 columns]

# Splitting the data into train and test

In [30]:
# split the predictors and target data into training and testing datasets
x_train,x_test,y_train,y_test = train_test_split(X,y,test_size = 0.3)

In [31]:
#understanding the size and structure of your training and testing data
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((356, 28), (356,), (153, 28), (153,))

# inferenece
the training dataset consists of 356 samples with their corresponding features and target values, while the testing dataset consists of 153 samples with their corresponding features and target values.

In [32]:
#creates and trains a Support Vector Machine
model_linear = SVC(kernel = "linear")
model_linear.fit(x_train,y_train)

SVC(kernel='linear')

In [33]:
#stores the predicted values in the variable pred_test_linear
pred_test_linear = model_linear.predict(x_test)

In [34]:
np.mean(pred_test_linear==y_test) # Accuracy = 98.46%

0.9869281045751634

# inference
 the SVM model achieved an accuracy of approximately 98.69% on the test data, meaning it correctly classified 98.69% of the test samples.

In [35]:
# calculates the accuracy of the SVM model's predictions
acc = accuracy_score(y_test, pred_test_linear) * 100
print("Accuracy =", acc)
confusion_matrix(y_test, pred_test_linear)

Accuracy = 98.69281045751634


array([[115,   2],
       [  0,  36]], dtype=int64)

# inference
The output shows the accuracy of the SVM model's predictions as 98.69, which means the model correctly classified 98.69% of the test samples. Additionally, the confusion matrix is displayed as an array with two rows and two columns. Each row represents the actual class, while each column represents the predicted class. The values in the array indicate the counts of samples falling into different categories. In this case, the confusion matrix shows that there were 36 true negatives, 2 false positive, 0 false negatives, and 115 true positives in the predictions made by the SVM model.

In [36]:
# Kernel = poly
model_poly = SVC(kernel = "poly")
model_poly.fit(x_train,y_train)

SVC(kernel='poly')

In [37]:
pred_test_poly = model_poly.predict(x_test)

In [38]:
#calculates the proportion of correct predictions
np.mean(pred_test_poly==y_test)

0.8235294117647058

# inference
The resulting value 0.8235 indicates that the SVM model with a polynomial kernel achieved a proportion of approximately 0.8235 (or 82.35%) correct predictions on the test data.

In [42]:
# kernel = rbf
model_rbf = SVC(kernel = "rbf")
model_rbf.fit(x_train,y_train)

SVC()

In [43]:
pred_test_rbf = model_rbf.predict(x_test)

In [44]:
#calculates the proportion of correct predictions
np.mean(pred_test_rbf==y_test) 

0.7908496732026143

# inference
The resulting value 0.7908 indicates that the SVM model with an RBF kernel achieved a proportion of approximately 0.7908 (or 79.08%) correct predictions on the test data.

In [45]:
#'sigmoid'
model_sig = SVC(kernel = "sigmoid")
model_sig.fit(x_train,y_train)

SVC(kernel='sigmoid')

In [46]:
pred_test_sig = model_rbf.predict(x_test)

In [47]:
np.mean(pred_test_sig==y_test)

0.7908496732026143

# #inference
The resulting value 0.79084 indicates that the SVM model with a sigmoid kernel achieved a proportion of approximately 0.7908 (or 79.08%) correct predictions on the test data.